flink-runtime/src/main/java/org/apache/flink/runtime/operators/chaining/ChainedAllGroupReduceCombineDriver.java

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.runtime.operators.chaining;
+
+import java.util.ArrayList;
+
+import org.apache.flink.api.common.functions.FlatCombineFunction;
+import org.apache.flink.api.common.functions.Function;
+import org.apache.flink.api.common.functions.util.FunctionUtils;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.TypeSerializerFactory;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable;
+import org.apache.flink.runtime.operators.RegularPactTask;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ChainedAllGroupReduceCombineDriver<T> extends ChainedDriver<T, T> {
+	private static final Logger LOG = LoggerFactory.getLogger(ChainedAllGroupReduceCombineDriver.class);
+
+	// --------------------------------------------------------------------------------------------
+	private FlatCombineFunction<T> combiner;
+	private TypeSerializer<T> serializer;
+
+	private volatile boolean running = true;
+
+	private final ArrayList<T> values = new ArrayList();
+
+	// --------------------------------------------------------------------------------------------
+	@Override
+	public void setup(AbstractInvokable parent) {
+		@SuppressWarnings("unchecked")
+		final FlatCombineFunction<T> com = RegularPactTask.instantiateUserCode(this.config, userCodeClassLoader, FlatCombineFunction.class);
+		this.combiner = com;
+		FunctionUtils.setFunctionRuntimeContext(com, getUdfRuntimeContext());
+
+		this.objectReuseEnabled = this.executionConfig.isObjectReuseEnabled();
+
+		final TypeSerializerFactory<T> serializerFactory = this.config.getInputSerializer(0, this.userCodeClassLoader);
+		this.serializer = serializerFactory.getSerializer();
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("ChainedAllGroupReduceCombineDriver object reuse: " + (this.objectReuseEnabled ? "ENABLED" : "DISABLED") + ".");
+		}
+	}
+
+	@Override
+	public void openTask() throws Exception {
+		final Configuration stubConfig = this.config.getStubParameters();
+		RegularPactTask.openUserCode(this.combiner, stubConfig);
+	}
+
+	@Override
+	public void closeTask() throws Exception {
+		if (!this.running) {
+			return;
+		}
+		RegularPactTask.closeUserCode(this.combiner);
+	}
+
+	@Override
+	public void cancelTask() {
+		this.running = false;
+	}
+
+	// --------------------------------------------------------------------------------------------
+	@Override
+	public Function getStub() {
+		return this.combiner;
+	}
+
+	@Override
+	public String getTaskName() {
+		return this.taskName;
+	}
+
+	// --------------------------------------------------------------------------------------------
+	@Override
+	public void collect(T record) {
+		try {
+			values.add(objectReuseEnabled ? record : serializer.copy(record));
+			if (values.size() > 10000) {
```

Comment:
is there a more reasonable value for this? (basically anything whose reasoning goes beyond "i felt like using 10000")